### DISCUSSION TITLE
0.33.0

### DISCUSSION THREAD

**Discussion by twyatt** [JuulLabs/kable#696] [Releases]
> [!IMPORTANT]
> Kable's Maven **artifact ID**s have changed (prefixed with `kable-`):
>
> ```kotlin
> kotlin {
>   sourceSets {
>     commonMain.dependencies {
>       implementation("com.juul.kable:kable-core:${kableVersion}")
>       implementation("com.juul.kable:kable-exceptions:${kableVersion}")
>       implementation("com.juul.kable:kable-log-engine-khronicle:${kableVersion}")
>     }
>   }
> }
> ```
> _Kable's package names were not changed, only the Maven coordinates; you should only need to update your Gradle configuration (e.g. `libs.versions.toml`)._

## Kotlin 2.0.0 (K2) Upgrade

Kable is now built against Kotlin 2.0.0 (K2). 🎉 

> [!NOTE]
> With this upgrade, the Kotlin compiler became more restrictive with `expect`/`actual` declarations, as a result, some Kable APIs have changed:
>
> - Advertisement interfaces (e.g. `AndroidAdvertisement`, `CoreBluetoothAdvertisement`, etc) have been consolidated to a common `PlatformAdvertisement`
> - `Scanner` interface now defines a generic type for advertisements it will emit: `Scanner<PlatformAdvertisement>`
>
> _For most users, these changes should not impact your code. Some notable instances where you'll need to make code changes:_
>
> If you hold a reference to a `Scanner` in a variable or property, you'll need to change `Scanner` to use the `PlatformScanner` type alias, or specify the generic type explicitly, for example:
>
> ```kotlin
> val scanner: PlatformScanner // or...
> val scanner: Scanner<PlatformAdvertisement>
> ```
>
> If you have any classes that implement the `Scanner` interface, you'll need to change the interface to `PlatformScanner` (or `Scanner<PlatformAdvertisement>`) and change the overridden type of the `advertisements` property to `PlatformAdvertisement`, for example:
>
> ```kotlin
> class ExampleScanner(..) : PlatformScanner {
>     override val advertisements: Flow<PlatformAdvertisement> = ..
> }
> ```

## 🚀 Changes

- Update dependency org.jetbrains.kotlin.multiplatform to v2 (#683)
- Replace Tuulbox logging w/ Khronicle (#690)
- Prefix Maven artifacts with `kable-` (#692)
- Remove `log-engine-tuulbox` module (#689)

## 🧰 Maintenance

- Update plugin android-library to v8.5.0 (#693)
- Update tuulbox to v8 (major) (#687)
- Update dependency gradle to v8.8 (#691)
- Update tuulbox to v7.3.0 (#686)
- Update Gradle documentation w/ new Maven coordinate (#694)

<hr /><em>This discussion was created from the release <a href='https://github.com/JuulLabs/kable/releases/tag/0.33.0'>0.33.0</a>.</em>

**Comment by JanusCole** [comment]
Hello, can I ask what your versioning strategy is and, specifically, if 0.33.0 is considered an alpha, beta, or production release?

**Comment by twyatt** [comment]
The versioning follows [semver](https://semver.org/) and is an indication of API stability (the versioning is not intended to give any indication of how production ready / bug-free the code is).

That being said, internally, we use all targets extensively, with us using Android and JS targets in production; and it has been incredibly stable.

Although, it is not without bugs. The [currently known bugs](https://github.com/JuulLabs/kable/issues?q=is%3Aissue+is%3Aopen+label%3Abug) are either rare or only occur in edge-cases.

As for "why pre-1.x versioning": the API is expected to change quite a bit, for example, the following issues give hints as to where the API is heading: #724, #737, #724, #286

The major pain point currently is that exceptions are not unified across platforms, making using Kable from `common` code difficult/error-prone. There is an active large refactor currently in progress to address this and many of the above mentioned issues.

I would say that using pre-1.x of Kable means that you'll likely have to make a number of code changes as the Kable API evolves and you upgrade to the newer versions.
